design/design-doc-packages-v2.md

```diff
@@ -0,0 +1,1387 @@
+
+1. Observe delete of `pkg-a`
+2. Get `Module` and check if `pkg-a` exists (Y)
+3. Check if `pkg-a` is depended on by any other packages (N)
+4. Check if any `PackageRevision`s exist (Y)
+5. Delete all `PackageRevision`s for `Package`
+6. Return
+
+*Reconcile 2 - Remove Finalizer*
+
+1. Observe delete of `pkg-a`
+2. Get `Module` and check if `pkg-a` exists (Y)
+3. Check if `pkg-a` is depended on by any other packages (N)
+4. Check if any `PackageRevision`s exist (N)
+  - If a `PackageRevision` has been deleted but is failing to clean up, that
+    status should be propagated to `Package`
+5. Remove finalizer from `Package`
+6. Return
+
+Note that deleting a parent package had no effect on the packages that we
+installed as its dependencies. Furthermore, the intermediate packages (B, C, D)
+are still able to resolve the correct version that is to be used for
+`provider-aws` and `provider-gcp`. Users in the cluster could continue to
+interact with each package in any way they were before. Let's now consider if
+`pkg-a` had not been deleted, but a user tried to delete
+`registry.upbound.io-crossplane-pkg-b`.
+
+*Reconcile 1 - Observe Delete*
+
+1. Observe delete of `registry.upbound.io-crossplane-pkg-b`
+2. Get `Module` and check if `registry.upbound.io-crossplane-pkg-b` exists (Y)
+3. Check if `registry.upbound.io-crossplane-pkg-b` is depended on by any other
+   packages (Y)
+4. Do not proceed and propagate invalid operation to status of
+   `registry.upbound.io-crossplane-pkg-b`
+6. Return
+
+For more scenarios that are required to be supported by the dependency
+management model, see the [Scenarios](#scenarios) section.
+
+Action Items
+
+- [ ] Explore the limitations of including all dependency information in a
+  single instance of an object (i.e. `Module`)
+  - [ ] How cumbersome would it be to "overflow" to other `Module` instances and
+    build a single pane on reconcile?
+  - [ ] Could `Module` state be an interface that allows for different backends
+    other than an object in `etcd` to be used?
+
+### Installation
+
+Once a `PackageRevision` has been created, the _`Package` controller_ hands
+control to the _`PackageRevision` controller_. The _`PackageRevision`
+controller_ will operate similarly to the current implementation the _`Package`
+controller_, with a few modifications. In total, the _`PackageRevision`_
+controller will execute the following steps:
+
+*`PackageRevision` Observed with State `Active`*
+
+1. Observe `PackageRevision` with state `Active`.
+3. Create all objects listed in `spec.customResourceDefinitions`,
+   `spec.infrastructureDefinitions`, `spec.InfrastructurePublications`, and
+   `spec.compositions`.
+4. Label the objects to indicate that they are currently owned by this
+   `PackageRevision`.
+5. Ensure that all dependent (listed in `spec.dependsOn`) CRDs are present in
+   the cluster.
+   - This should have been checked by the _`Package` controller_, but we check
+     here again.
```

Comment:
What's the thinking behind doing this again?

Reply:
This can be removed now because with the `Module` type we are deferring all unpack and install to the `PackageRevision` controller. Previously we were saying that we would make sure that the CRDs were present, then the  `PackageRevision` would just establish ownership and start its packaged controller. Now, the package has "reserved" the CRDs in the `Module` so the `PackageRevision` controller is free to carry out its operations as needed, which could include creating / adding labels / etc to any CRDs or composition types, then starting the packaged controller (if one was provided).